codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java

```diff
@@ -354,4 +285,34 @@ public ByteString path() {
     public ByteString status() {
         return get(PseudoHeaderName.STATUS.value());
     }
+
+    private static class Http2HeaderNameComparator implements Comparator<ByteString>, Serializable {
+
+        private static final Http2HeaderNameComparator INSTANCE = new Http2HeaderNameComparator();
+        private static final long serialVersionUID = 1109871697664666478L;
+
+        @Override
+        public int compare(ByteString one, ByteString two) {
+            // Reserved header names come first.
+            final boolean isPseudoHeader1 = one.byteAt(0) == ':';
+            final boolean isPseudoHeader2 = two.byteAt(0) == ':';
+            if (isPseudoHeader1 && !isPseudoHeader2) {
+                return -1;
+            }
+            if (!isPseudoHeader1 && isPseudoHeader2) {
+                return 1;
+            }
+            final int delta = one.hashCode() - two.hashCode();
+            if (delta == 0) {
+                // If the hash code matches it's very likely for the two strings to be equal
+                // and thus we optimistically compare them with the much faster equals method.
+                if (one.equals(two)) {
```

Comment:
ByteString has a [DEFAULT_COMPARATOR](https://github.com/netty/netty/blob/master/common/src/main/java/io/netty/util/ByteString.java#L39) which we could optimize.  `ByteString` didn't implement `Comparable<ByteString>` because `AsciiString` already implemented `Comparable<CharSequence>` and then java would be confused if you had an `AsciiString` and tried to compare it against another `AsciiString`...I'm open for suggestions on this one.